### ISSUE TITLE
Terraform Cloud and cloud backend on workspace creation based on tag

### ISSUE THREAD

**Issue by zopar** [hashicorp/terraform#31322]
I do not think this is a bug but more a design that I would like to invite to reconsider.

When you use backend cloud on TF Cloud with cli-driven workflow is possible to create workspace based on tags or with name.

Problem 1
If you use name, you can create only one workspace, how this will fit with repository that implements more than one environment and a CI system?
Because in the cloud backed you can assign only one name.
backend-config and related variables does not works in cloud backend, so you can pass the backend based on the environment.

Problem 2
If you decide to use tag, you can create "automatically" workspaces via cli and this is useful in a CI system.
The problematic part is related to this implementation:
https://github.com/hashicorp/terraform/blob/1c9929a3780104b060ba57a1b5a931485ed83141/internal/command/meta_backend.go#L217

First of all, it is mandatory to run a terraform init
If you do not have any workspace, terraform cli asks to write a name to create one. It is the first workspace and the init want an answer is not possible to do that in a CI pipeline. 
Because it is not possible to create the first workspace with an option in cli or something similar. The only workaround is to use a command like this

echo "first_workspace" | terraform init

Also If a workspace already exists, terraform cli automatically use it. In a pipeline this is really dangerous. Because the system decide to default to a workspace whit all related consequences if it is the wrong workspace.

This works on the first run of a CI BUT on a second run, when you need a new workspace, not works because terraform initialize the first found workspace. In a pipeline this is really dangerous. Because the system decide to default to a workspace whit all related consequences if it is the wrong workspace.

Finally you need the second one, you need to do 
terraform workspace new second_workspace
that is completely unrelated to the terraform init that create the first workspace or default to the first one.

All these things over complicating a CI system and the logic of use of the cli. Also is error prone and we know that an error on a workspace can be destructive.

Is it possible to think to a better design?
Maybe the best one could be the possibility to add multiple workspace in the backend. On the first init it will create all the workspace without any default and the user use terraform workspace select.
So it will be transparent about what happens.

Another possibility is the ability to create the first_workspace with a paramenter like "create if does not exists"

Another one in the cloud backend to setup a variable that will be populated based on the environment and with the related init that permit to initialize without problem the workspace.

Another one  permit the use of backend-config so during the init is possible to initialize the workspace based on an environment.

In general we can have multiple approach. What we have now  is probably the worst case scenario and should be improved.

Meanwhile, do you have any suggestion or workaround? Also about implementing the creation of first and others workspace in a pipeline (please do not tell me about api call). More in general about automatic workspace creation for multiple environments.

**Comment by apparentlymart** [comment]
Hi @zopar,

Since this is a Terraform Cloud-related request I have labelled it to help the Terraform Cloud teams find it, since ultimately they will be the ones deciding how best to respond.

You asked for suggestions for how you can achieve this today, but then you immediately said not to suggest the thing that is usually suggested: if you want to interact with Terraform Cloud in automation, the primary way to do that is via its API. The CLI workflow is intended for interactive use, and wraps the Terraform Cloud API in a way that is convenient for human users to operate, including interactive prompts when necessary.

My only other suggestions would be variations on that same idea, such as using [the `hashicorp/tfe` provider](https://registry.terraform.io/providers/hashicorp/tfe/0.32.1) to manage Terraform Cloud workspaces using the normal Terraform plan/apply process. Again that is just another wrapper around the same API, but maybe it's easier for you to wrap in some automation.

**Comment by zopar** [comment]
@apparentlymart Thank you for your answer.
The reason why I asked to not suggest to use API is that we can risk to spend more time to interact via API with terraform cloud than to build our infrastracture. 

A part of that I want also to try API, could you please suggest some repo examples that implement it? A small workflow that permit to simply plan and apply via API from a CI system, like gitlab for example. I will be happy to read it and explore this possibility.

A small modification on the cloud provider can help many people to use cli-driven workflow that in many case is enough to use.

For example a simple different exit status on terraform init can help to distinguish between when it fails because no workspace found and you need to create the first one and when fails because you have already 2 workspaces and you need to select one.

Thank you for your kind reply.

**Comment by apparentlymart** [comment]
In your question it seemed like you were primarily concerned with creating workspaces and so I was imagining using the [Create a Workspace](https://www.terraform.io/cloud-docs/api-docs/workspaces#create-a-workspace) API to do that -- or equivalently to use the `hashicorp/tfe` provider to manage them indirectly with Terraform -- and then that would avoid the situation you described where the workspace you intend to use doesn't exist yet, while allowing you to still use the CLI workflow for the main apply step if you want.

Starting a new run via the API _is_ a considerably more involved process because there are [multiple steps to create the context needed to start the run](https://www.terraform.io/cloud-docs/run/api). But I think you should be able to use Terraform CLI to drive that part without any interaction, once you've created the workspace(s) you intend to use, aside from the limitation that you won't be able to save a plan to disk using `terraform plan -out=tfplan` and then apply it separately using `terraform apply tfplan`, because in Terraform Cloud it is the remote system which saves the plan and then applies it separately and so the local client does not have access to the plan file.

**Comment by zopar** [comment]
First of all thank you for your answer. I will try for sure about api.
I think also that with some improvement on terraform cli the majority of user can use terraform cloud without investigating more in the API. I try to explain.

1) Terraform cli can create automatically workspaces. The only fix needed is in the output/exit code related to init, one ask for the name of new workspace, one to select one of the existing workspace. This can be bypassed with an if inside the script pipelines but having a different exist status is the minimum to better automate this part.
This type of workspace is in general enough, obviously if you need more is bettere to use tfe or for simple mod some api calls.

2) In an integration with an external pipeline, could be useful to have the plan out, otherwise is necessary to find a way to download it from TFC to reuse inside the pipeline for viewing the speculative plan.

3) The "confusion" happens because in remote executor the terraform apply is not like the old style apply, this create a plan and after asks if you want to apply. This behavior completely changes how a pipeline can works with cli-driven workflow.
Probably with a pipeline we should need to run terraform apply 2 times, whan to take the "real" plan and not the speculative one and another one to apply it.

Basically with a cli-driven is necessary to rethink about a pipeline that integrate with terraform cloud or simply end to use the local executor, but at this point we lost other advantages of terraform cloud.

Regarding API it is also more complicated. For example here the repo with some non production script that show how to use TFC with API https://github.com/hashicorp/terraform-guides/tree/master/operations/automation-script This approach could needs too much effort. With TFE will be for sure better.
